### PR TITLE
Log loss in WandB and other fixes

### PR DESCRIPTION
**Pull request by alejandromarcu**
The main thing in this PR is to log the loss in WandB so we easily see how it evolves.  I ended up adding a lot of other changes, some related some completely orthogonal. 
The loss track ended up being tricky because we were using "steps" but you can only increase them monotonically and they need to be int, and I needed a different way to track losses.  I find it very confusing the way wandb does it, but I defined the metrics "Loss step" and "Episode" which would be the x axis:

```
        wandb.define_metric("Loss step", hidden=True)
        wandb.define_metric("Episode", hidden=True)
```

And then I indicated that all the metrics starting with loss_ will use 'Loss Step' and the rest 'Episode`:
```
        wandb.define_metric("loss_*", "Loss step")
        wandb.define_metric("*", "Episode")
```

E.g. from [this run](https://wandb.ai/the-lazy-learning-lair/deep_quoridor/runs/amarcu-20250926-000648?nw=nwuseramarcu)

image

The training for each epoch is in the X coordinates between that epoch number and extends for 0.5, so we can distinguish epochs.

E.g. when zooming in:
image
Between 0 and 0.5 I see how the training evolved after epoch 0.  Then from 1 to 1.5 how it evolved in epoch 1 and so on.

### CODE REVIEW

File: deep_quoridor/src/agents/alphazero/alphazero.py [hunk 1]
```diff
@@ -491,7 +503,7 @@ def save_replay_buffer_to_file(self, episode_number: int):
 
     def load_replay_buffer_from_file(self) -> bool:
         """Load replay buffer from file if it exists. Returns True if the file exists."""
-        if self.params.train_every is None:
+        if self.params.train_every is None or not self.is_training():
```
Comment (alejandromarcu) [inline deep_quoridor/src/agents/alphazero/alphazero.py:506]: It was trying to load the replay buffer when training and causing an error

File: deep_quoridor/src/agents/alphazero/nn_evaluator.py [hunk 2]
```diff
@@ -187,6 +188,9 @@ def split_data(self, replay_buffer, validation_ratio: float) -> tuple[list, list
         return training_set, validation_set
 
     def compute_losses(self, batch_data):
+        if len(batch_data) == 0:
+            return None, None, None
```
Comment (alejandromarcu) [inline deep_quoridor/src/agents/alphazero/nn_evaluator.py:192]: this makes simpler the code that calls it

File: deep_quoridor/src/agents/alphazero/nn_evaluator.py [hunk 3]
```diff
@@ -221,44 +225,69 @@ def compute_losses(self, batch_data):
...
-        return {"total_loss": total_loss.item(), "policy_loss": policy_loss.item(), "value_loss": value_loss.item()}
+        # Get a last entry for the losses after the last backprop
+        with torch.no_grad():
+            batch_data = random.sample(training_set, self.batch_size)
+            policy_loss, value_loss, total_loss = self.compute_losses(batch_data)
+            v_policy_loss, v_value_loss, v_total_loss = self.compute_losses(validation_set)
+            entry = make_entry(i + 1, total_loss, policy_loss, value_loss, v_total_loss, v_policy_loss, v_value_loss)
+
+            if on_new_entry is not None:
+                on_new_entry(entry)
```
Comment (alejandromarcu) [inline deep_quoridor/src/agents/alphazero/nn_evaluator.py:291]: We were not computing the loss after the last training step, but I think it's good to have it as well, and allows us to see how the loss changes when we have different data in the next epoch

File: deep_quoridor/src/agents/alphazero/self_play_manager.py [hunk 4]
```diff
@@ -64,6 +63,14 @@ def run(self):
...
             self._run_central_evaluation()
 
+    # Return an array with the number of games per worker, such that they add up to the total number of games
+    # and each worker gets X or X+1 jobs.
+    # E.g. for 20 games and 8 workers it returns: [3, 3, 3, 3, 2, 2, 2, 2]
+    def _games_per_worker(self):
+        base = self.num_games // self.num_workers
+        extra = self.num_games - base * self.num_workers
+        return [base + 1 if i < extra else base for i in range(self.num_workers)]
+
```
Comment (alejandromarcu) [inline deep_quoridor/src/agents/alphazero/self_play_manager.py:73]: I noticed when playing 20 games with 8 workers that it was assigning 3 per worker, so it was playing 24 games.
Then I realized it was missing `games_remaining_to_allocate -= this_worker_num_games`, but this didn't work as I expected, since in that case it was doing 6 workers with 3 games, 1 worker with 2 and another worker empty.

File: deep_quoridor/src/metrics.py [hunk 5]
```diff
@@ -12,11 +12,20 @@ class Metrics:
...
+        benchmarks: list[str | Agent] = [],
+        benchmarks_t: int = 10,
+        max_steps=200,
+    ):
         self.board_size = board_size
         self.max_walls = max_walls
         self.stored_elos = {}
         self.benchmarks = benchmarks
+        self.benchmarks_t = benchmarks_t
+        self.max_steps = max_steps
```
Comment (alejandromarcu) [inline deep_quoridor/src/metrics.py:28]: This ones were hardcoded before, and I made it so that you can specify from the arguments now

File: deep_quoridor/src/plugins/wandb_train.py [hunk 6]
```diff
@@ -148,23 +147,19 @@ def compute_tournament_metrics(self, model_filename: str) -> int:
             self.best_model_filename = model_filename
             print("Best Relative Elo so far!")
 
-        wandb_elo_table = wandb.Table(
-            columns=["Player", "elo"], data=[[player, elo] for player, elo in elo_table.items()]
-        )
```
Comment (alejandromarcu) [inline deep_quoridor/src/plugins/wandb_train.py:153]: this table was useless, so just removing it

File: deep_quoridor/src/quoridor.py [hunk 7]
```diff
@@ -60,7 +60,7 @@ def __init__(self, board_size: int):
         self.board_size = board_size
         self.wall_size = board_size - 1
         self.num_actions = self.board_size**2 + self.wall_size**2 * 2
-        self.constant_action_mask_template = np.zeros(self.num_actions, dtype=np.float32)
+        self.constant_action_mask_template = np.zeros(self.num_actions, dtype=np.int8)
```
Comment (alejandromarcu) [inline deep_quoridor/src/quoridor.py:63]: When playing against greedy:p_random=0.5, I was getting the following error:

```
  File "/Users/amarcu/code/deep_rabbit_hole/deep_quoridor/src/metrics.py", line 130, in compute
    results = arena._play_games(players, self.benchmarks_t, PlayMode.ALL_VS_ALL)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/amarcu/code/deep_rabbit_hole/deep_quoridor/src/arena.py", line 180, in _play_games
    result = self._play_game(agent_1, agent_2, f"game_{match_id:04d}")
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/amarcu/code/deep_rabbit_hole/deep_quoridor/src/arena.py", line 94, in _play_game
    action = int(agent.get_action(observation_before_action))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/amarcu/code/deep_rabbit_hole/deep_quoridor/src/agents/greedy.py", line 185, in get_action
    return self.action_space.sample(action_mask)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/amarcu/code/deep_rabbit_hole/.venv/lib/python3.12/site-packages/gymnasium/spaces/discrete.py", line 93, in sample
    mask.dtype == np.int8
```

So I just changed it to be int8 as requested.  Notice that if you are running with -O the assertion is disabled, it took me a while to understand why when training it was not giving that error (I was copy pasting a training with -O, but I was running something else without -O)

File: deep_quoridor/src/train_alphazero.py [hunk 8]
```diff
@@ -46,6 +47,12 @@ def train_alphazero(
         # members, so we pass in a GameParams object. We have to call start_game
         # because it calls the plugin's internal _intialize method which sets up metrics.
         wandb_train_plugin.start_game(game=args, agent1=training_agent, agent2=training_agent)
+        training_agent.set_wandb_run(wandb_train_plugin.run)
+
+        # Compute the tournament metrics with the initial model, possibly random initialized, to
+        # be able to see how it evolves from there
+        wandb_train_plugin.episode_count = 0
+        wandb_train_plugin.compute_tournament_metrics(str(current_filename))
```
Comment (alejandromarcu) [inline deep_quoridor/src/train_alphazero.py:55]: The untrained model can already beat some agents during MCTS, so it's good to have it as a base

File: deep_quoridor/src/agents/alphazero/self_play_manager.py (hunk 4 above)
Comment (jonbinney) [reply deep_quoridor/src/agents/alphazero/self_play_manager.py:73]: I definitely need to do a better job of debugging before making PR's :-( Sorry!

File: deep_quoridor/src/metrics.py (hunk 5 above)
Comment (jonbinney) [reply deep_quoridor/src/metrics.py:28]: Ah, this is great! I've been wishing that the number of games in the benchmarks was configurable!

File: deep_quoridor/src/quoridor.py (hunk 7 above)
Comment (jonbinney) [reply deep_quoridor/src/quoridor.py:63]: I think that @muralx used float32 here to make the vector easier to do math with during training? But I think it makes sense as an int8 if that's what PettingZoo expects.

**Review by jonbinney — APPROVED** [review]
LGTM! Lots of good changes.

File: deep_quoridor/src/agents/alphazero/alphazero.py [hunk 9]
```diff
@@ -438,12 +448,14 @@ def train_iteration(self, is_replay_buffer_bootstrap=False) -> bool:
...
+            print("  Epoch       Total   Val Total   Policy  Val Policy  Value   Val Value")
+        else:
+            print("==== Training Loss ====")
+            print("  Epoch   Total   Policy  Value")
 
-        # Store loss for metrics
-        self.recent_losses.append(metrics["total_loss"])
-        if len(self.recent_losses) > 100:  # Keep only recent losses
-            self.recent_losses = self.recent_losses[-100:]
+        self.evaluator.train_iteration(self.replay_buffer, self.params.validation_ratio, on_new_entry=log_loss_entry)
```
[the remark this answers is not in the document]
Comment (alejandromarcu) [reply deep_quoridor/src/agents/alphazero/alphazero.py:458]: We could consider making it into a generator (i.e. a function that uses yield to give control to the caller, like range).
Then we would do something like:
```
for entry in evaluator.train_step(num_batches=100):
    # entry will have the losses for the step
```
